src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonSerializer
+    {
+        private static void VerifyValueAndType(object value, Type type)
+        {
+            if (type == null)
+            {
+                if (value != null)
+                {
+                    throw new ArgumentNullException(nameof(type));
+                }
+            }
+            else if (value != null)
+            {
+                if (!type.IsAssignableFrom(value.GetType()))
+                {
+                    ThrowHelper.ThrowArgumentException_DeserializeWrongType(type, value);
+                }
+            }
+        }
+
+        private static void WriteNull(
+            ref JsonWriterState writerState,
+            IBufferWriter<byte> bufferWriter)
+        {
+            Utf8JsonWriter writer = new Utf8JsonWriter(bufferWriter, writerState);
+            writer.WriteNullValue();
+            writer.Flush(true);
+        }
+
+        private static byte[] WriteCore(object value, Type type, JsonSerializerOptions options)
```

Comment:
This should return `Memory<byte>` instead (or even a `Span<byte>`) and the caller should call `.ToArray()` on it where needed.
Otherwise, we are unnecessarily allocating for writing to string:
```C#
// BEFORE:
        private static string ToStringInternal(object value, Type type, JsonSerializerOptions options)
        {
            Span<byte> jsonBytes = WriteCore(value, type, options); // Allocates a byte array
            string stringJson = JsonReaderHelper.TranscodeHelper(jsonBytes); // Allocates a string
            return stringJson;
        }
//AFTER:
        private static string ToStringInternal(object value, Type type, JsonSerializerOptions options)
        {
            Memory<byte> jsonBytes = WriteCore(value, type, options); // No allocations
            string stringJson = JsonReaderHelper.TranscodeHelper(jsonBytes.Span); // Allocates a string
            return stringJson;
        }
```

Reply:
`WriteCore` returns `byte[]` that was copied from a pooled array which is now disposed. So I'll refactor the code to call the `TranscodeHelper` from the `WriteCore` method where the Memory<> is available.